# Mobile App Market Analysis: Maximising User Engagement and Ad Revenue

## Introduction
This project analyses data from Android and iOS mobile apps available on Google Play and the App Store. The focus is on exploring app datasets to uncover trends and patterns within the mobile app market, from the perspective of a data analyst working for a company that builds free-to-download mobile apps.

## The Goal
The primary source of revenue is in-app ads, meaning profitability is directly tied to user volume. The main goal of this project is to determine what types of apps are most likely to attract a large user base. These data-driven insights will inform app development decisions aimed at maximising user engagement and ad revenue.

## Data Sources
- A dataset containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018. Download directly from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv).
- A dataset containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017. Download directly from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv).

## Opening and Exploring the Data

In [1]:
from csv import reader

# Google Play Dataset
open_file = open('googleplaystore.csv')
read_file = reader(open_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

# Apple Store Dataset
open_file = open('AppleStore.csv')
read_file = reader(open_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

The helper function `explore_data()` below prints dataset rows in a readable format, with an optional parameter to display the total number of rows and columns.

In [2]:
def explore_data(dataset,start,end,rows_and_columns = False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')

    if rows_and_columns:
        print('Numbers of rows:', len(dataset))
        print('Numbers of columns:', len(dataset[0]))        

In [3]:
# Explore Google Play dataset

print(android_header)
print('\n')
explore_data(android,1,3,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Numbers of rows: 10841
Numbers of columns: 13


In [4]:
# Explore Apple Store dataset

print(ios_header)
print('\n')
explore_data(ios,1,3,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Numbers of rows: 7197
Numbers of columns: 16


## Data Cleaning

### Removing Rows with Missing Columns

The Google Play dataset has a dedicated [discussion section](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion?sort=undefined), and [one of the discussions](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472. The row is printed below and compared against the header.

#### Option 1 - .index() function

Python's built-in index() function is a useful tool for finding the index of a specific element in a sequence. This function takes an argument representing the value to search for and returns the index of the **first occurrence** of that value in the sequence.

In [5]:
for row in android:
    if len(row) != len(android_header):
        index = android.index(row)
        print(f"Index: {index}") # f-string, which is a clean way to insert variables directly into a text string.
        print(f"Row Data: {row}")
        print('\n')

Index: 10472
Row Data: ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']




Using .index() inside a loop can be highly inefficient for large datasets. Because .index() searches through the dataset from the beginning each time to find the match, it significantly increases the computational time.

Also, the .index() function always returns the index of the **very first match** it finds. If the dataset happens to have duplicate rows, and row 10472 is identical to row 500, .index() will return 500, even though the loop is currently looking at row 10472!

#### Option 2 - enumerate() function

A much more efficient approach is to use Python's built-in enumerate() function. This provides both the index and the row data simultaneously as you iterate, completely bypassing the need to scan the list for the index separately.

In [6]:
for index, row in enumerate(android):
    if len(row) != len(android_header):
        print(f"Row {index} has missing columns")
        print(f"Correct total columns: {len(android_header)}")
        print(f"Total columns in row {index}: {len(row)}")
        print('\n')
        print(f"header: {android_header}")
        print('\n')
        print(f"Row {index}: {row}")
        

Row 10472 has missing columns
Correct total columns: 13
Total columns in row 10472: 12


header: ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Row 10472: ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [7]:
# Delete row 10,472
print(len(android))
del android[10472]  # Don't run this more than once!
print(len(android))

10841
10840


In [8]:
# Check Apple Store dataset for rows with missing columns

# Create the tracker flag BEFORE the loop starts
missing_found = False

for index, row in enumerate(ios):
    if len(row) != len(ios_header):
        print(index)
        print(row)
        missing_found = True # This only triggers if a bad row exists

# Check the flag after the loop is done    
if missing_found == False:
    print(f"There's no row with missing columns")

There's no row with missing columns


### Removing Duplicate Entries

The Google Play dataset contains apps with multiple entries. These duplicates fall into two categories:

- Updated data: apps like Instagram have multiple entries because they were scraped at different times, resulting in varying review counts
- Identical copies: apps like Box have multiple entries that are exactly the same across every column

In [9]:
print(android_header)

print('\n')

for app in android:
    name = app[0]
    if name == "Instagram":
        print(app)

print('\n')

for app in android:
    name = app[0]
    if name == "Box":
        print(app)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with dev

In [10]:
# Identify duplicate apps in the Google Play dataset

duplicate_google_apps = []
unique_google_apps = []

for app in android:
    name = app[0]
    if name in unique_google_apps:
        duplicate_google_apps.append(name)
    else:
        unique_google_apps.append(name)

print("Number of duplicate apps:", len(duplicate_google_apps))
print('\n')
print("Example of duplicate apps:", duplicate_google_apps[:10])


Number of duplicate apps: 1181


Example of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


The Google Play dataset contains 1,181 duplicate entries. Rather than removing duplicates randomly, the row with the highest number of reviews is kept for each app. The higher the review count, the more recent and reliable the data.

The approach taken:

- Build a dictionary where each key is a unique app name and the value is the highest review count found for that app
- Use the dictionary to create a cleaned dataset containing only one entry per app, keeping the row with the highest review count


In [11]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])

    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        

There are 1,181 cases where an app occurs more than once, so the length of the dictionary of unique apps should equal the difference between the total dataset length and 1,181.

In [12]:
print("Expected row:", len(android)-1181)
print(len(reviews_max))

Expected row: 9659
9659


The `reviews_max` dictionary is used to remove duplicates, keeping only the entry with the highest review count for each app.

The logic:

- Two empty lists are initialised: `android_clean` and `already_added`
- For each app in the dataset, the name and review count are extracted
- The row is added to `android_clean` only if its review count matches the maximum recorded in `reviews_max` and the app name has not already been added
- The second condition handles cases where a duplicate app has identical review counts across all entries, such as Box, which would otherwise still produce duplicates

In [13]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])

    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name) # make sure this is inside the if block

A quick check confirms the cleaned dataset contains 9,659 rows, as expected.

In [14]:
explore_data(android_clean,0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Numbers of rows: 9659
Numbers of columns: 13


In [15]:
# Identify duplicate apps in the Apple Store dataset

duplicate_ios_apps = []
unique_ios_apps = []

for app in ios:
    name = app[0]
    if name in unique_ios_apps:
        duplicate_ios_apps.append(name)
    else:
        unique_ios_apps.append(name)

print("Number of duplicate apps:", len(duplicate_ios_apps))
print('\n')
print("Example of duplicate apps:", duplicate_ios_apps[:10])

Number of duplicate apps: 0


Example of duplicate apps: []


No duplicate apps in the Apple Store dataset.

### Removing Non-English Apps

Several non-English apps are present in both datasets. Since this analysis focuses on the English-speaking market, these entries are removed.

In [16]:
# Some examples of noon-English apps
print(ios[813][1])
print(ios[6731][1])

print('\n')

print(android_clean[4412][0])
print(android_clean[7940][0])


爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
لعبة تقدر تربح DZ


One way to identify non-English apps is to use Python's [built-in ord() function](https://docs.python.org/3/library/functions.html#ord). According to the [ASCII system](https://en.wikipedia.org/wiki/ASCII), common English characters fall within the range of 0 to 127. A function can be built to check each character in an app name against this range. Any character with a value above 127 is considered non-English.

**For context**, English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

In [17]:
# Create a function to determine if a string is English. 

def is_english(string):
    for character in string:
        if ord(character) > 127:
            return False
            
    # 'return True' must sit here at the for-else level, not inside the if block.
    # Placing it inside would cause the function to return True on the very first
    # English character, exiting before any non-English character is ever checked.
    # e.g. "Instagr汉m" would incorrectly return True on the first letter "I".
    else: 
        return True

In [18]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

False
False


The function works, but some legitimate English app names use emojis or special symbols (™, em dash, en dash, etc.) that fall outside the ASCII range. In its current form, the function would incorrectly remove these apps.

To minimise data loss, an app is only removed if its name contains **more than three** non-ASCII characters:


In [19]:
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
        

True
True


The filter is not perfect. A small number of non-English apps may still pass through, but this level of precision is sufficient for the purposes of this analysis.

The `is_english()` function is applied below to filter both datasets:

In [20]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)

for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)


explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Numbers of rows: 9614
Numbers of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+'

After filtering, 9,614 Android apps and 6,183 iOS apps remain.

### Isolating the Free Apps

Since the business model relies on in-app ads, only free apps are relevant to this analysis. Both datasets contain paid apps, which are removed below. The price column is at index 7 for Android and index 4 for iOS.

In [21]:
# Print headers to identify the index of the Price column in each dataset
print(android_header) 
print('\n')
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [22]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if (price =='0') or (price == '0.0'):
        android_final.append(app)
        
for app in ios_english:
    price = app[4]
    if (price =='0') or (price == '0.0') :
        ios_final.append(app)

print(len(android_final))
print(len(ios_final))

8864
3222


After isolating free apps, 8,864 Android apps and 3,222 iOS apps remain, which is sufficient for the analysis.


## Data Analysis

The data cleaning steps completed:

- Removed inaccurate data
- Removed duplicate app entries
- Removed non-English apps
- Isolated free apps

The goal of this analysis is to determine the kinds of apps most likely to attract a large user base, since user volume directly affects ad revenue.

The validation strategy for a new app idea follows three steps:

1. Build a minimal Android version and publish it to Google Play
2. If user response is positive, develop it further
3. If the app is profitable after six months, build an iOS version for the App Store

Since the target is success on both platforms, the analysis looks for app profiles that perform well in both markets.

### Most Common Apps by Genre

The analysis starts by identifying the most common genres in each market. A frequency table is built for the `prime_genre` column of the App Store dataset, and the `Genres` and `Category` columns of the Google Play dataset.

In [23]:
# ================================================
# FUNCTION 1: Frequency Generator
# Calculates the percentage of apps in each genre
# ================================================
def freq_table(dataset, index):
    table = {} # Empty dictionary to store the raw frequency counts
    total = 0  # Accumulator to track the total number of rows

    # 1. Aggregate raw frequencies
    for row in dataset:
        total += 1             # Increment the total for each row
        value = row[index] # Extract the target variable (e.g., genre)
        # Populate the dictionary with counts
        if value in table:
            table[value] += 1
        else: 
            table[value] = 1

    # 2. Compute percentages
    table_percentages = {}
    for key in table:
        # Calculate the relative frequency as a percentage
        percentage = (table[key]/total) *100
        table_percentages[key] = percentage

    return table_percentages

# =========================================================
# FUNCTION 2: Display Output
# Sorts and prints the frequency table in descending order
# =========================================================
def display_table(dataset, index):
    # Retrieve the computed percentages
    table = freq_table(dataset, index)

    # Convert the dictionary into a list of tuples for sorting
    table_display = []
    for key in table:
        # Structure as (percentage, category_name) to sort numerically
        key_val_as_tuple = (table[key], key) #Tuple here
        table_display.append(key_val_as_tuple)

    # Sort the list of tuples in descending order
    table_sorted = sorted(table_display, reverse = True)

    # Print the sorted distribution
    for entries in table_sorted:
        print(entries[1] , ":", entries[0])

**Why use a tuple instead of a dictionary or a list?**

**1. Why not just sort the dictionary?**
Dictionaries are built for speed and key-based lookups, not for maintaining order. Passing a dictionary directly to Python's sorted() function will result in Python ignoring the values (the percentages) and only sorting the keys (the genre names) in alphabetical order. Because the objective is to sort the data numerically by percentage, the dictionary format is insufficient.

**2. Why a tuple instead of a list?** Technically, structuring the data as a list of lists ([percentage, genre]) instead of a list of tuples ((percentage, genre)) would still work, as Python sorts both data types by evaluating the first item. However, data scientists use tuples in this scenario to communicate **intention**.

Lists are mutable; they are designed for dynamic data that may be appended, removed, or reorganised later. Tuples are immutable and are meant for fixed records. Because a specific genre is permanently locked to its specific percentage in this dataset, a tuple is the safest and most "Pythonic" way to package them together. It clearly signals that this pair is a finalised record that should not be altered.

#### Apple Store Dataset
Examining the frequency table for the `prime_genre` column

In [24]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Among free English apps, more than half (58.16%) are games. Entertainment apps account for close to 8%, followed by photo and video apps at nearly 5%. Education apps represent only 3.66%, and social networking apps 3.29%.

The App Store, at least the segment containing free English apps, is dominated by apps designed for entertainment (games, entertainment, photo and video, social networking, sports, music), while practical apps (education, shopping, utilities, productivity, lifestyle) are comparatively rare. However, the number of apps in a genre does not necessarily reflect the number of users. Supply and demand may not align.

#### Google Play dataset
Examining the `Genres` and `Category` columns

In [25]:
display_table(android_final, 1) # Category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The Google Play landscape looks significantly different. Apps designed purely for entertainment are less prominent. However, the Family category, which accounts for almost 19% of apps, consists mostly of games for children.

Practical apps have stronger representation on Google Play compared to the App Store. This is confirmed by the frequency table for the `Genres` column:

In [26]:
display_table(android_final, -4) #Genres

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The distinction between the Genres and Category columns is not entirely clear, but the Genres column is noticeably more granular. For a high-level view, the Category column is used going forward.

At this stage, the App Store appears dominated by entertainment-focused apps, while Google Play shows a more balanced mix of practical and entertainment apps. The next step is to identify which genres attract the most users.

### Most Popular Apps by Genre on the App Store



One way to identify the most popular genres is to calculate the **average number of installs** per genre. The Google Play dataset includes an `Installs` column, but this information is absent from the App Store dataset. The total number of user ratings (`rating_count_tot`) is used as a proxy.

#### Apple Store Dataset


In [27]:
# Get all unique genres from the iOS dataset (column -5 = genre column)
genre_ios = freq_table(ios_final, -5)

# Empty dictionary to store results
# Need this dictionary ONLY needed for sorting at the end, not during calculation

# Step 1: Calculate average ratings per genre
avg_ratings_by_genre = {}

for genre in genre_ios:
    total = 0       # Reset running total for this genre
    len_genre = 0   # Reset app counter for this genre
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre: # Check if this app belongs to the genre currently being checked
            n_rating = float(app[5])
            total += n_rating
            len_genre += 1
    avg_n_rating = total / len_genre
    avg_ratings_by_genre[genre] = avg_n_rating # Store result in dictionary e.g. {'Games': 22788.67, 'Music': 57326.53}
    
# Sort by value (highest to lowest) and print
# .items()           → returns (genre, number) pairs from the dictionary
# key=lambda x: x[1] → sort by number (position 1), not genre name (position 0)
# reverse=True       → highest first
for genre, avg_n_rating in sorted(avg_ratings_by_genre.items(), key = lambda x: x[1], reverse = True):
        print(genre, ":", avg_n_rating)

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Food & Drink : 33333.92307692308
Finance : 31467.944444444445
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22788.6696905016
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 18684.456790123455
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


On average, navigation apps have the highest number of user ratings, but this figure is heavily skewed by Waze and Google Maps, which together account for close to half a million reviews.

In [28]:
for app in ios_final: 
    if app[-5] == "Navigation":
        print(app[1], ":", app[5]) # Print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social networking apps, where the average is skewed by giants like Facebook, Pinterest, and Skype, and to music apps, where Pandora, Spotify, and Shazam dominate.

These genres may appear more popular than they actually are. The averages are pulled up by a small number of apps with hundreds of thousands of ratings, while most other apps in the same genre struggle to reach 10,000. A more accurate picture could be obtained by removing these outliers before recalculating, which is left for a future iteration.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [29]:
for app in ios_final: 
    if app[-5] == "Reference":
        print(app[1], ":", app[5]) # Print name and number of ratings

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


Despite the skew, this genre shows potential. One opportunity is to build an app centred around a popular book, adding features beyond the raw text such as daily quotes, an audio version, chapter quizzes, or an embedded dictionary. This removes the need for users to switch to an external app for lookups.

This concept fits the App Store context well. The market is saturated with entertainment apps, which means a practical, content-rich app has a stronger chance of standing out.

Other genres with relatively high averages include weather, food and drink, and finance. The book genre overlaps with the concept above, but the remaining genres present structural challenges:

- Weather apps tend to have low in-app engagement time, which limits ad revenue potential. Reliable live weather data also typically requires paid third-party APIs.
- Food and drink apps are dominated by established brands with physical operations such as Starbucks and McDonald's, which is outside the scope of an app-only business.
- Finance apps cover banking, bill payments, and money transfers, all of which require specialist domain knowledge and regulatory awareness.

#### Google Play Dataset


The Google Play dataset includes install data, which gives a more direct measure of genre popularity. However, the values are open-ended (100+, 1,000+, 5,000+, etc.) and are not precise:

In [30]:
display_table(android_clean, 5)

1,000,000+ : 14.670255720053834
100,000+ : 11.522931980536287
10,000+ : 10.663629775339063
10,000,000+ : 9.700797183973496
1,000+ : 9.183145253131794
100+ : 7.340304379335334
5,000,000+ : 6.284294440418263
500,000+ : 5.217931462884357
50,000+ : 4.855575111295165
5,000+ : 4.8452220726783315
10+ : 3.985919867481106
500+ : 3.3957966663215657
50,000,000+ : 2.112019877834144
50+ : 2.112019877834144
100,000,000+ : 1.9567242985816338
5+ : 0.8489491665803914
1+ : 0.6936535873278807
500,000,000+ : 0.24847292680401697
1,000,000,000+ : 0.20706077233668083
0+ : 0.14494254063567658
0 : 0.01035303861683404


The install numbers are open-ended (100+, 1,000+, 5,000+, etc.), so exact figures are unavailable. For this analysis, each value is treated as its lower bound. An app with 100,000+ installs is counted as 100,000, and an app with 1,000,000+ installs is counted as 1,000,000.

Before computation, the comma and plus characters are stripped from each value and the result is converted to float. This is handled directly in the loop below, where the average number of installs per genre is also calculated.

In [31]:
genre_android = freq_table(android_final, 1)

avg_n_installs_by_genre = {}

for genre in genre_android:
    total = 0
    len_genre = 0
    for app in android_final:
        genre_app = app[1]
        if genre_app == genre:
            n_install = app[5]
            n_install = n_install.replace(",","")
            n_install = n_install.replace("+","")
            total += float(n_install)
            len_genre += 1
    
    avg_n_installs = total / len_genre
    avg_n_installs_by_genre[genre] = avg_n_installs

for genre, avg_n_installs in sorted(avg_n_installs_by_genre.items(), key = lambda x: x[1], reverse = True):
    print(genre, ":", avg_n_installs)

COMMUNICATION : 38456119.167247385
VIDEO_PLAYERS : 24727872.452830188
SOCIAL : 23253652.127118643
PHOTOGRAPHY : 17840110.40229885
PRODUCTIVITY : 16787331.344927534
GAME : 15588015.603248259
TRAVEL_AND_LOCAL : 13984077.710144928
ENTERTAINMENT : 11640705.88235294
TOOLS : 10801391.298666667
NEWS_AND_MAGAZINES : 9549178.467741935
BOOKS_AND_REFERENCE : 8767811.894736841
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
WEATHER : 5074486.197183099
HEALTH_AND_FITNESS : 4188821.9853479853
MAPS_AND_NAVIGATION : 4056941.7741935486
FAMILY : 3695641.8198090694
SPORTS : 3638640.1428571427
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.145631068
BUSINESS : 1712290.1474201474
LIFESTYLE : 1437816.2687861272
FINANCE : 1387692.475609756
HOUSE_AND_HOME : 1331540.5616438356
DATING : 854028.8303030303
COMICS : 817657.2727272727
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
PARENTING : 542603.6206896552
BEAUTY : 51315

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [32]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

Removing communication apps with over 100 million installs reduces the average roughly tenfold:

In [33]:
under_100m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(",","")
    n_installs = n_installs.replace("+","")
    if (app[1] == "COMMUNICATION") and (float(n_installs) < 100000000):
        under_100m.append(float(n_installs))

sum(under_100m) / len(under_100m)  

3603485.3884615386

The same skew pattern appears in the video players category, where YouTube and MX Player dominate, in social apps (Facebook, Instagram), photography apps (Google Photos), and productivity apps (Microsoft Word, Dropbox, Evernote).

The core concern across these categories is that a small number of dominant apps inflate the genre averages, making these niches appear more accessible than they are.

The game genre shows high average installs but was identified as saturated in the frequency analysis. The books and reference genre is a more interesting candidate, averaging 8,767,811 installs and showing potential on the App Store as well, making it a strong option for a cross-platform recommendation.

The apps in this genre and their install counts are listed below:

In [34]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:



In [35]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


Only a handful of apps dominate the top install counts, which suggests the market still has room. The apps sitting in the middle range, between 1,000,000 and 100,000,000 downloads, are examined below:

In [36]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche is largely occupied by ebook readers, library collections, and dictionary apps, all of which represent categories with significant existing competition.

The number of Quran-related apps with strong install counts suggests that building an app around a specific popular book is a viable approach. Taking a widely-read book and building a feature-rich app around it shows potential for both the Google Play and App Store markets.

The opportunity lies in going beyond a plain text version. Features such as daily quotes, an audio version, chapter quizzes, or a community discussion forum could differentiate a new entry in this space.

## Conclusion

Data from the App Store and Google Play was analysed to identify app profiles likely to be profitable across both platforms.

The books and reference genre shows consistent potential in both markets. The recommended direction is to build an app centred around a popular book, enhanced with features beyond the raw text such as daily quotes, an audio version, chapter quizzes, or an in-app discussion forum. This profile avoids the saturation seen in games and communication apps, while targeting a genre with demonstrated user demand across both platforms.